go/vt/wrangler/reparent.go

```diff
@@ -996,55 +995,279 @@ func (wr *Wrangler) emergencyReparentShardLocked(ctx context.Context, ev *events
+			}
+			cleanedGTIDSet := relayLogGTIDSet.Difference(errantGTIDs)
+			cleanedRelayLogPosition := mysql.Position{GTIDSet: cleanedGTIDSet}
+			positionMap[alias] = cleanedRelayLogPosition
+		} else {
+			positionMap[alias] = status.FileRelayLogPosition
+		}
+	}
+
+	for alias, masterStatus := range masterStatusMap {
+		if *gtidBased {
+			executedPosition, err := mysql.DecodePosition(masterStatus.Position)
+			if err != nil {
+				return nil, err
+			}
+			positionMap[alias] = executedPosition
+		} else {
+			executedFilePosition, err := mysql.DecodePosition(masterStatus.FilePosition)
+			if err != nil {
+				return nil, err
+			}
+			positionMap[alias] = executedFilePosition
+		}
+	}
+
+	if !*gtidBased {
+		// If all positions are not comparable, we can't do file based comparisons at all.
+		filePositionList := make([]mysql.Position, 0, len(replicationStatusMap))
+		for _, p := range positionMap {
+			filePositionList = append(filePositionList, p)
+		}
+		if !mysql.AllPositionsComparable(filePositionList) {
+			return nil, fmt.Errorf("we can't compare all file based positions")
+		}
+	}
+
+	// Find a position that is fully caught up when errant GTIDs are taken out of the equation.
+	var winningPosition mysql.Position
+	for _, position := range positionMap {
+		if winningPosition.IsZero() {
+			winningPosition = position
+		}
+		if position.AtLeast(winningPosition) {
+			winningPosition = position
+		}
+	}
+
+	// Now that we have a high water mark, we can get a list of all tablets that are errant GTID free
+	// that hit this high water mark.
+	var validCandidates []string
+	for alias, position := range positionMap {
+		if badTablets.Has(alias) {
+			// We should exclude tablets we've already determined have errant GTIDs from consideration.
+			continue
+		}
+
+		if position.AtLeast(winningPosition) {
+			validCandidates = append(validCandidates, alias)
+		}
+	}
+
+	return validCandidates, nil
+}
+
+func (wr *Wrangler) stopReplicationAndBuildStatusMaps(ctx context.Context, ev *events.Reparent, tabletMap map[string]*topo.TabletInfo, waitReplicasTimeout time.Duration, unreachableReplicas sets.String) (map[string]*replicationdatapb.StopReplicationStatus, map[string]*replicationdatapb.MasterStatus, error) {
+	// Stop replication on all replicas, get their current
+	// replication position
+	event.DispatchUpdate(ev, "stop replication on all replicas")
+	statusMap := make(map[string]*replicationdatapb.StopReplicationStatus)
+	masterStatusMap := make(map[string]*replicationdatapb.MasterStatus)
+	mu := sync.Mutex{}
+
+	errChan := make(chan error)
+	groupCtx, groupCancel := context.WithTimeout(ctx, waitReplicasTimeout)
+	defer groupCancel()
+	fillStatus := func(alias string, tabletInfo *topo.TabletInfo) {
+		err := fmt.Errorf("fillStatus did not successfully complete")
+		returnErr := func() { errChan <- err }
+		defer returnErr()
+
+		wr.logger.Infof("getting replication position from %v", alias)
+		ctx, cancel := context.WithCancel(groupCtx)
```

Comment:
I don't see any place this subcontext is cancelled while anyone could be using it, so it seems unnecessary to make a new subcontext.

Reply:
I had pulled this pattern over from the pre-existing code. I had assumed the previous offer did this to defer canceling of each goroutines call to `StopReplicationAndGetStatus`. That code used waitgroups. Am I missing some understanding of why the previous author had used this pattern?

Reply:
Agree with @enisoc, it should be fine to use groupCtx here.

Reply:
Switched to groupCtx